### PR TITLE
Feat/export import

### PR DESCRIPTION
**Pull request by CharlieS1103**

### CODE REVIEW

**Comment by CharlieS1103** [comment]
Making this pull request as a reminder that we should probably get this done.

**Comment by Tetrax-10** [comment]
you still gonna implement this and add it to marketplace? we have spotify backup.

**Comment by CharlieS1103** [comment]
Yes

**Comment by theRealPadster** [comment]
This isn't done...

**Comment by CharlieS1103** [comment]
@theRealPadster Yeah I did get it working though, essentially all we need to do is make it look nice and it'd work. Although, we may want to transition to loading in the imports through a file rather than the text-area (for the sake of speed.)

**Comment by theRealPadster** [comment]
> @theRealPadster Yeah I did get it working though, essentially all we need to do is make it look nice and it'd work. Although, we may want to transition to loading in the imports through a file rather than the text-area (for the sake of speed.)

I'm taking a look now and though it does copy to clipboard fine, pasting a massive amount of data back in still freezes the client. I think using a file is the only way. Idk how to make it download a file though. My method I was testing with making an `<a>` tag with the `download` attribute set and the `href` set to a blob URI didn't seem to do anything. 

We should be able to use a file input to upload and parse though, since I got that working for the snippet images.

**Comment by CharlieS1103** [comment]
We don't need to download as a file if copying works( probably just impossible), and yeah we can just upload as file

**Comment by theRealPadster** [comment]
> We don't need to download as a file if copying works( probably just impossible), and yeah we can just upload as file

Yeah don't need to, but it would make it easier for people to transfer and back things up if they don't need to paste it into a text editor and save their own.

**Comment by CharlieS1103** [comment]
> > We don't need to download as a file if copying works( probably just impossible), and yeah we can just upload as file
> 
> 
> 
> Yeah don't need to, but it would make it easier for people to transfer and back things up if they don't need to paste it into a text editor and save their own. 

Yeah you're definitely right but if I had to guess, it will be impossible. Might give it a try regardless though.

File: packages/marketplace/src/components/Modals/ImportExport/index.tsx [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    // Check if pastedData is valid JSON, if not return an error message and exit
+    try {
+      settings = JSON.parse(pastedData);
+    } catch (e) {
+      Spicetify.showNotification("Invalid JSON");
+      return;
+    }
+    importMarketplace(settings);
+    Spicetify.PopupModal.hide();
+    location.reload();
```
Comment (theRealPadster) [inline packages/marketplace/src/components/Modals/ImportExport/index.tsx:43]: Yeah if we're doing a reload here, we don't need it in the import function itself

File: packages/marketplace/src/components/Modals/ImportExport/index.tsx [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            placeholder="Copy/paste your settings here"
+            style={{
+              // fontFamily: "'Fira code', 'Fira Mono', monospace'",
+              // fontSize: 12,
+            }}
+          />
+        </div>
+      </div>
+      <>
+        {/* <a href={href} className="download" download="marketplace-export.json">
```
Comment (theRealPadster) [inline packages/marketplace/src/components/Modals/ImportExport/index.tsx:91]: Remove if we can't get the download to work

File: packages/marketplace/src/components/Modals/ImportExport/index.tsx [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    catch (e) {
+      Spicetify.showNotification("Invalid JSON");
+      return;
+    }
+    importMarketplace(settings);
+
+  };
+
+  return (
+    // TODO: remove "add-snippet" and "customCSS" references
```
Comment (theRealPadster) [inline packages/marketplace/src/components/Modals/ImportExport/index.tsx:69]: Remove these references. Probably make class names more generic

File: packages/marketplace/src/components/Modals/ImportExport/index.tsx [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+import "prismjs/components/prism-json";
+
+import {
+  exportMarketplace,
+  importMarketplace,
+} from "../../../logic/Utils";
+import Button from "../../Button";
+
+const ImportExportModal = () => {
+  const [code, setCode] = React.useState("");
```
Comment (theRealPadster) [inline packages/marketplace/src/components/Modals/ImportExport/index.tsx:13]: Name it something more accurate, like settings

File: packages/marketplace/package.json [hunk 5]
```diff
@@ -43,5 +43,9 @@
 		"spicetify-creator": "^1.0.11",
 		"typescript": "^4.8.4"
 	},
-	"private": true
+	"private": true,
+	"dependencies": {
```
Comment (theRealPadster) [inline packages/marketplace/package.json:47]: These should be dev dependencies

**Pull request merged**